### ISSUE TITLE
Embedding lookup table doesn't mask padding value

### ISSUE THREAD

**Issue by aleSuglia** [tensorflow/tensorflow#2373]
Hi there,

I'm using an *embedding_lookup* operation in order to generate dense vector representations for each token in my document which are feed to a convolutional neural network (the network architecture is similar to the one in a [WildML article](http://www.wildml.com/2015/12/implementing-a-cnn-for-text-classification-in-tensorflow/)). 

Everything works correctly but when I pad my document inserting a padding value in it, the embedding lookup generates a vector for this token too. I think that this approach could alterate the results in the classification task. What I want to achieve is something similar to what Torch [LookupTableMaskZero](https://github.com/Element-Research/rnn#rnn.LookupTableMaskZerol) does. 

1) Is correct what I want to do?
2) Is already implemented something like this? 
3) If not, how can I mask the padding value in order to prevent the generation of the corresponding vector for it? 

Thank you in advance, 
Alessandro

**Comment by LittleYUYU** [comment]
I guess I met the same problem, @aleSuglia . 

In the released "tranlsate" codes of a seq2seq model, buckets are used for defining different lengths of encoders and decoders. Every time we get a batch of data, we pad the original input with "PAD" to satisfy the required length of a bucket, e.g., [1,2,3] --> [1,2,3,PAD,PAD] for the bucket with length=5. However, we take the PAD as a normal token in embedding. That's strange.

In training a seq2seq model, we usually reverse the input sequence to be [PAD, PAD, 3, 2, 1]. In this case, the embedding of PAD can change the final decoding result. In my experiments, when testing new data without knowing its corresponding bucket, the decoding result would be different with different buckets.

So I guess that's a bug? Thanks much!

Ziyu

**Comment by aleSuglia** [comment]
Exactly @LittleYUYU! 

I don't know why nobody has noticed it before but for me is a missing feature and not a bug. I hope that somebody, one day, will notice this issue and fix it. 

Thank you to all in advance.

Alessandro Suglia

**Comment by LittleYUYU** [comment]
@aleSuglia Right! Currently I use a mask which is a matrix with the same dimension as the embedding matrix but with the padding value set to be zero to solve this problem. I just feel confused: why the released sample code doesn't consider this issue? Thanks!

Ziyu

**Comment by girving** [comment]
I think the sample code doesn't consider the issue because it doesn't affect neural network performance.  If you're seeing significant differences between zero padding an new-token padding, I would use a mask as @LittleYUYU suggests.  `embedding_lookup` doesn't handle this by itself since we try to keep the semantics of each op simple and focused.

**Comment by shawnjhenry** [comment]
This can actually cause a fairly substantial fluctuation in performance in some networks.  Suppose that instead of a convnet, we feed the embeddings to a deep averaging network.  Then the varying number of nonzero pad vectors (according to which training batch the example is assigned in SGD) will very much affect the value of the average embedding.  I've seen this cause a variation of up to 3% accuracy on text classification tasks.  One possible remedy is to set the value of the embedding for the pad index to zero explicitly between each backprop step during training.  This is computationally kind of wasteful (and also requires explicitly feeding the number of nonzero embeddings in each sample to the network), but it does the trick.  I would like to see a feature like Torch's LookupTableMaskZero as well.

**Comment by sonalgupta** [comment]
I would also be interested in a feature like this. Both Theano and Keras support it.

**Comment by buptjz** [comment]
Has this problem been solved yet?

**Comment by memeda** [comment]
Hi, I come across this problem and get into this issue by search. I'm so glad that this question is proposed properly, which indicating a SEEK-THE-PRECISE attitude!

I'm not very familiar with TF, so @LittleYUYU 's methods description still un-clearly. SO I come across the code at [Embedding lookup table doesn't mask padding value](http://stackoverflow.com/questions/37255038/embedding-lookup-table-doesnt-mask-padding-value) (the same question in stackoverflow), I rewrite it to support arbitrary PADDING ID instead only 0.

```Python
# omit some CONSTANT
raw_mask_array = [[1.]] * PADDING_ID + [[0.]] + [[1.]] * (WORDS_NUM - PADDING_ID - 1)
with tf.variable_scope("Embedding"):
    lookup_table = tf.get_variable("lookup_table", shape=[WORDS_NUM, EMBEDDING_DIM],
                                   initializer=tf.random_uniform_initializer(
                                               minval=-R, maxval=R),
                                   dtype=DTYPE,
                                   trainable=True)
    mask_padding_lookup_table = tf.get_variable("mask_padding_lookup_table",
                                                initializer=raw_mask_array,
                                                dtype=DTYPE,
                                                trainable=False)
embedding_input = tf.nn.embedding_lookup(lookup_table, id_input)
mask_padding_input = tf.nn.embedding_lookup(mask_padding_lookup_table, id_input)
embedding_input = tf.multiply(embedding_input, mask_padding_input) # broadcast
``` 
what's more, I make a more simple solution(which has been write down in that stackoverflow page):

```Python
mask_padding_zero_op = tf.scatter_update(lookup_table, 
                                         PADDING_ID, 
                                         tf.zeros([EMBEDDING_DIM,], dtype=DTYPE))
# just for explicity
lookup_table = mask_padding_zero_op
```

just to reset PADDING ID's embedding value to zeros and replace it. see the stackoverflow page for details.

this code may be naive or not property, writing here for some  people like me searching this. Any errors please point out.

**Comment by youngia** [comment]
@memeda it returned "No gradient defined for operation 'embedding/ScatterUpdate" when use tf.scatter_update, seems it can't work with embedding variable as trainable

**Comment by memeda** [comment]
@youngia , thanks you for response. I really have not use it in practice. I'll delete it 😭

**Comment by julienbelan** [comment]
Is there anyway to mask the padding 0s then, I use to do it before embedding but it made my autoencoder lazy (obviously). I've heard that embedding twice with the second time with idx set to some 0 mask worked. Anybody heard of that?

**Comment by ay27** [comment]
@memeda @youngia , I follow the code but with slightly change and succeed:
```
mask_padding_zero_op = tf.scatter_update(lookup_table, 
                                         PADDING_ID, 
                                         tf.zeros([EMBEDDING_DIM,], dtype=DTYPE))
with tf.control_dependencies([mask_padding_zero_op]):
    # do embedding lookup...
```

**Comment by newtiopon** [comment]
@ay27 i tried but still get the error under tf1.12.0
